File: libs/search-abstraction/src/opensearch.rs
//! OpenSearch-backed [`SearchBackend`].
//!
//! OpenSearch is the dev / CI fallback per
//! [ADR-0028](../../docs/architecture/adr/ADR-0028-search-backend-abstraction.md).
//!
//! Wire format:
//!
//! * Index per `(tenant, type_id)` named `of-{tenant}-{type}`
//!   (sanitized, lowercase).
//! * `PUT  /{index}/_doc/{id}?if_seq_no=…&if_primary_term=…` — we
//!   use the `version` field with `version_type=external` so stale
//!   writes are silently dropped (HTTP 409).
//! * `POST /{index}/_search` for both lexical (`query_string`) and
//!   vector (`knn`) queries.
//! * `POST /_bulk` for `bulk_index`.
//!
//! We deliberately do **not** use the official `opensearch` crate.
//! Its dependency footprint (aws-sigv4, hyper-tls, …) is heavier
//! than the surface we exercise; raw `reqwest` against the REST API
//! keeps the build small and the wire format auditable in code
//! review.

use async_trait::async_trait;
use serde_json::{Value, json};

use crate::{
    BulkOutcome, IndexDoc, ObjectId, PagedResult, ReadConsistency, RepoError, RepoResult,
    SearchBackend, SearchHit, SearchQuery, TenantId, TypeId, VectorQuery, sanitize_doc_type,
};

/// OpenSearch client. Construct with the cluster's HTTP endpoint
/// (typically `http://opensearch:9200` in `compose dev`).
pub struct OpenSearchBackend {
    endpoint: String,
    http: reqwest::Client,
}

impl OpenSearchBackend {
    /// Build with an explicit endpoint and a default `reqwest::Client`.
    pub fn new(endpoint: impl Into<String>) -> Self {
        Self::with_client(endpoint, reqwest::Client::new())
    }

    /// Build with a caller-provided HTTP client.
    pub fn with_client(endpoint: impl Into<String>, http: reqwest::Client) -> Self {
        let mut endpoint = endpoint.into();
        if endpoint.ends_with('/') {
            endpoint.pop();
        }
        Self { endpoint, http }
    }

    fn index_name(&self, tenant: &TenantId, type_id: &TypeId) -> String {
        format!(
            "of-{}-{}",
            sanitize_index(&tenant.0),
            sanitize_doc_type(&type_id.0)
        )
    }

    fn doc_url(&self, index: &str, id: &ObjectId) -> String {
        format!("{}/{}/_doc/{}", self.endpoint, index, urlencode(&id.0))
    }
}

fn sanitize_index(s: &str) -> String {
    s.to_ascii_lowercase()
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '-' || c == '_' {
                c
            } else {
                '_'
            }
        })
        .collect()
}

fn urlencode(s: &str) -> String {
    s.bytes()
        .map(|b| match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                (b as char).to_string()
            }
            other => format!("%{:02X}", other),
        })
        .collect()
}

fn map_status(status: reqwest::StatusCode, ctx: &'static str) -> Option<RepoError> {
    if status.is_success() {
        None
    } else {
        Some(RepoError::Backend(format!(
            "opensearch {ctx} HTTP {}: {}",
            status.as_u16(),
            status.canonical_reason().unwrap_or("")
        )))
    }
}

#[async_trait]
impl SearchBackend for OpenSearchBackend {
    async fn search(
        &self,
        query: SearchQuery,
        _consistency: ReadConsistency,
    ) -> RepoResult<PagedResult<SearchHit>> {
        // Pick index: when `type_id` is unspecified we hit the
        // tenant-wide alias (`of-{tenant}-*`); when present we
        // target the concrete index for a tighter shard scan.
        let target = match &query.type_id {
            Some(t) => self.index_name(&query.tenant, t),
            None => format!("of-{}-*", sanitize_index(&query.tenant.0)),
        };
        let url = format!("{}/{}/_search", self.endpoint, target);

        let mut filter: Vec<Value> = vec![json!({"term": {"tenant": query.tenant.0}})];
        if let Some(t) = &query.type_id {
            filter.push(json!({"term": {"type_id": t.0}}));
        }
        for (k, v) in &query.filters {
            filter.push(json!({"term": {k: v}}));
        }

        let mut must: Vec<Value> = Vec::new();
        if let Some(q) = &query.q {
            if !q.is_empty() {
                must.push(json!({"query_string": {"query": q}}));
            }
        }
        if must.is_empty() {
            must.push(json!({"match_all": {}}));
        }

        let body = json!({
            "size": query.page.size.max(1),
            "query": {"bool": {"must": must, "filter": filter}},
        });

        let resp = self
            .http
            .post(url)
            .json(&body)
            .send()
            .await
            .map_err(|e| RepoError::Backend(format!("opensearch search send: {e}")))?;
        if resp.status().as_u16() == 404 {
            return Ok(PagedResult {
                items: vec![],
                next_token: None,
            });
        }
        if let Some(err) = map_status(resp.status(), "search") {
            return Err(err);
        }
        let v: Value = resp
            .json()
            .await
            .map_err(|e| RepoError::Backend(format!("opensearch search decode: {e}")))?;
        Ok(PagedResult {
            items: parse_hits(&v),
            next_token: None,
        })
    }

    async fn index(&self, doc: IndexDoc) -> RepoResult<()> {
        let index = self.index_name(&doc.tenant, &doc.type_id);
        let url = self.doc_url(&index, &doc.id);
        let mut payload = match doc.payload.clone() {
            Value::Object(map) => Value::Object(map),
            other => json!({ "payload": other }),
        };
        if let Some(map) = payload.as_object_mut() {
            map.insert("id".into(), json!(doc.id.0));
            map.insert("tenant".into(), json!(doc.tenant.0));
            map.insert("type_id".into(), json!(doc.type_id.0));
            map.insert("version".into(), json!(doc.version));
            if let Some(emb) = &doc.embedding {
                map.insert("embedding".into(), json!(emb));
            }
        }

        let resp = self
            .http
            .put(url)
            .query(&[
                ("version", doc.version.to_string().as_str()),
                ("version_type", "external"),
            ])
            .json(&payload)
            .send()
            .await
            .map_err(|e| RepoError::Backend(format!("opensearch index send: {e}")))?;
        // 409 = stale (external version conflict) — silently drop.
        if resp.status().as_u16() == 409 {
            return Ok(());
        }
        if let Some(err) = map_status(resp.status(), "index") {
            return Err(err);
        }
        Ok(())
    }

    async fn delete(&self, tenant: &TenantId, id: &ObjectId) -> RepoResult<bool> {
        // Like Vespa, we don't know the type_id; use a wildcard
        // delete-by-query.
        let target = format!("of-{}-*", sanitize_index(&tenant.0));
        let url = format!("{}/{}/_delete_by_query", self.endpoint, target);
        let body = json!({
            "query": {
                "bool": {
                    "filter": [
                        {"term": {"tenant": tenant.0}},
                        {"term": {"id": id.0}},
                    ]
                }
            }
        });
        let resp = self
            .http
            .post(url)
            .json(&body)
            .send()
            .await
            .map_err(|e| RepoError::Backend(format!("opensearch delete send: {e}")))?;
        if resp.status().as_u16() == 404 {
            return Ok(false);
        }
        if let Some(err) = map_status(resp.status(), "delete") {
            return Err(err);
        }
        let v: Value = resp
            .json()
            .await
            .map_err(|e| RepoError::Backend(format!("opensearch delete decode: {e}")))?;
        let deleted = v.get("deleted").and_then(|x| x.as_u64()).unwrap_or(0);
        Ok(deleted > 0)
    }

    async fn search_vector(
        &self,
        query: VectorQuery,
        _consistency: ReadConsistency,
    ) -> RepoResult<Vec<SearchHit>> {
        let target = match &query.type_id {
            Some(t) => self.index_name(&query.tenant, t),
            None => format!("of-{}-*", sanitize_index(&query.tenant.0)),
        };
        let url = format!("{}/{}/_search", self.endpoint, target);
        let mut filter: Vec<Value> = vec![json!({"term": {"tenant": query.tenant.0}})];
        if let Some(t) = &query.type_id {
            filter.push(json!({"term": {"type_id": t.0}}));
        }
        for (k, v) in &query.filters {
            filter.push(json!({"term": {k: v}}));
        }
        let body = json!({
            "size": query.k.max(1),
            "query": {
                "knn": {
                    "embedding": {
                        "vector": query.embedding,
                        "k": query.k.max(1),
                        "filter": {"bool": {"filter": filter}}
                    }
                }
            }
        });
        let resp = self
            .http
            .post(url)
            .json(&body)
            .send()
            .await
            .map_err(|e| RepoError::Backend(format!("opensearch vector send: {e}")))?;
        if resp.status().as_u16() == 404 {
            return Ok(vec![]);
        }
        if let Some(err) = map_status(resp.status(), "search_vector") {
            return Err(err);
        }
        let v: Value = resp
            .json()
            .await
            .map_err(|e| RepoError::Backend(format!("opensearch vector decode: {e}")))?;
        Ok(parse_hits(&v))
    }

    async fn bulk_index(&self, docs: Vec<IndexDoc>) -> RepoResult<BulkOutcome> {
        if docs.is_empty() {
            return Ok(BulkOutcome::default());
        }
        let url = format!("{}/_bulk", self.endpoint);
        let mut body = String::new();
        for d in &docs {
            let index = self.index_name(&d.tenant, &d.type_id);
            let header = json!({
                "index": {
                    "_index": index,
                    "_id": d.id.0,
                    "version": d.version,
                    "version_type": "external",
                }
            });
            let mut payload = match d.payload.clone() {
                Value::Object(map) => Value::Object(map),
                other => json!({ "payload": other }),
            };
            if let Some(map) = payload.as_object_mut() {
                map.insert("id".into(), json!(d.id.0));
                map.insert("tenant".into(), json!(d.tenant.0));
                map.insert("type_id".into(), json!(d.type_id.0));
                map.insert("version".into(), json!(d.version));
                if let Some(emb) = &d.embedding {
                    map.insert("embedding".into(), json!(emb));
                }
            }
            body.push_str(&serde_json::to_string(&header).unwrap());
            body.push('\n');
            body.push_str(&serde_json::to_string(&payload).unwrap());
            body.push('\n');
        }
        let resp = self
            .http
            .post(url)
            .header("Content-Type", "application/x-ndjson")
            .body(body)
            .send()
            .await
            .map_err(|e| RepoError::Backend(format!("opensearch bulk send: {e}")))?;
        if let Some(err) = map_status(resp.status(), "bulk_index") {
            return Err(err);
        }
        let v: Value = resp
            .json()
            .await
            .map_err(|e| RepoError::Backend(format!("opensearch bulk decode: {e}")))?;
        let mut out = BulkOutcome::default();
        if let Some(items) = v.get("items").and_then(|x| x.as_array()) {
            for (i, entry) in items.iter().enumerate() {
                let result = entry.get("index").or_else(|| entry.get("create"));
                let status = result
                    .and_then(|r| r.get("status"))
                    .and_then(|s| s.as_u64());
                let id = docs
                    .get(i)
                    .map(|d| d.id.clone())
                    .unwrap_or(ObjectId(String::new()));
                match status {
                    Some(s) if (200..300).contains(&s) || s == 409 => out.indexed += 1,
                    Some(s) => out.failed.push((id, format!("status {s}"))),
                    None => out.failed.push((id, "missing status".into())),
                }
            }
        }
        Ok(out)
    }
}

fn parse_hits(v: &Value) -> Vec<SearchHit> {
    let Some(hits) = v
        .get("hits")
        .and_then(|h| h.get("hits"))
        .and_then(|h| h.as_array())
    else {
        return Vec::new();
    };
    hits.iter()
        .filter_map(|h| {
            let src = h.get("_source")?;
            let id = src
                .get("id")
                .and_then(|x| x.as_str())
                .or_else(|| h.get("_id").and_then(|x| x.as_str()))?
                .to_string();
            let type_id = src
                .get("type_id")
                .and_then(|x| x.as_str())
                .unwrap_or("")
                .to_string();
            let score = h.get("_score").and_then(|x| x.as_f64()).unwrap_or(0.0) as f32;
            Some(SearchHit {
                id: ObjectId(id),
                type_id: TypeId(type_id),
                score,
                snippet: Some(src.clone()),
            })
        })
        .collect()
}

File: libs/search-abstraction/opensearch/opensearch.go
// Package opensearch is the OpenSearch-backed SearchBackend
// implementation. Mirrors libs/search-abstraction/src/opensearch.rs.
//
// Wire format:
//
//   - Index per (tenant, type_id) named `of-{tenant}-{type}`
//     (sanitized, lowercase).
//   - PUT  /{index}/_doc/{id}?version=…&version_type=external —
//     stale writes are silently dropped (HTTP 409).
//   - POST /{index}/_search for both lexical (`query_string`) and
//     vector (`knn`) queries.
//   - POST /_bulk for BulkIndex.
//
// Importing this package as a side-effect registers the backend
// with searchabstraction.SearchBackendFromEnv:
//
//	import _ "github.com/openfoundry/openfoundry-go/libs/search-abstraction/opensearch"
//
// We deliberately do not depend on the official `opensearch-go`
// client. The trait surface we exercise is small (six methods) and
// the JSON wire format is stable; pulling in the official client
// would drag in `aws-sdk-go-v2/aws/signer` and a parallel HTTP
// stack we already have in `net/http`. Documented divergence from
// S0.8.c (Rust note carries over).
package opensearch

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	searchabstraction "github.com/openfoundry/openfoundry-go/libs/search-abstraction"
	repos "github.com/openfoundry/openfoundry-go/libs/storage-abstraction"
)

func init() {
	searchabstraction.RegisterBackend(searchabstraction.BackendOpenSearch,
		func(endpoint string) searchabstraction.SearchBackend { return New(endpoint) })
}

// Backend is the OpenSearch HTTP client. Construct with the cluster
// HTTP endpoint (typically `http://opensearch:9200` in `compose dev`).
type Backend struct {
	endpoint   string
	http       *http.Client
	authHeader string
}

// Option customises a Backend.
type Option func(*Backend)

// WithHTTPClient configures the HTTP client used by the backend.
func WithHTTPClient(client *http.Client) Option {
	return func(b *Backend) {
		if client != nil {
			b.http = client
		}
	}
}

// WithAuthHeader configures the Authorization header sent to OpenSearch.
func WithAuthHeader(value string) Option {
	return func(b *Backend) { b.authHeader = strings.TrimSpace(value) }
}

// New builds a Backend with a default *http.Client (30 s timeout).
func New(endpoint string) *Backend {
	return NewWithOptions(endpoint)
}

// NewWithOptions builds a Backend with optional transport/auth configuration.
func NewWithOptions(endpoint string, opts ...Option) *Backend {
	b := WithClient(endpoint, &http.Client{Timeout: 30 * time.Second})
	for _, opt := range opts {
		if opt != nil {
			opt(b)
		}
	}
	return b
}

// WithClient builds a Backend with a caller-provided *http.Client.
func WithClient(endpoint string, client *http.Client) *Backend {
	if client == nil {
		client = http.DefaultClient
	}
	return &Backend{endpoint: strings.TrimRight(endpoint, "/"), http: client}
}

// Compile-time interface satisfaction.
var _ searchabstraction.SearchBackend = (*Backend)(nil)

// ─── helpers ───────────────────────────────────────────────────────────

func (b *Backend) indexName(tenant repos.TenantId, typeID repos.TypeId) string {
	return fmt.Sprintf("of-%s-%s",
		sanitizeIndex(string(tenant)),
		searchabstraction.SanitizeDocType(string(typeID)))
}

func (b *Backend) docURL(index string, id repos.ObjectId) string {
	return fmt.Sprintf("%s/%s/_doc/%s", b.endpoint, index, percentEncode(string(id)))
}

func sanitizeIndex(s string) string {
	out := make([]byte, 0, len(s))
	for _, c := range strings.ToLower(s) {
		if (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '-' || c == '_' {
			out = append(out, byte(c))
		} else {
			out = append(out, '_')
		}
	}
	return string(out)
}

// percentEncode mirrors the Rust impl: preserve the unreserved set
// (RFC 3986: [A-Za-z0-9-_.~]) and percent-encode every other byte
// of the UTF-8 representation. Stays compatible with OpenSearch's
// path normaliser.
func percentEncode(s string) string {
	var w strings.Builder
	for _, c := range []byte(s) {
		switch {
		case (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') ||
			(c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' || c == '~':
			w.WriteByte(c)
		default:
			fmt.Fprintf(&w, "%%%02X", c)
		}
	}
	return w.String()
}

func mapStatus(status int, ctx string) error {
	if status >= 200 && status < 300 {
		return nil
	}
	return repos.Backend(fmt.Sprintf("opensearch %s HTTP %d: %s",
		ctx, status, http.StatusText(status)))
}

// augmentPayload is the shared serialiser for Index + BulkIndex.
// Mirrors the Rust pattern: if the payload is a JSON object, mutate
// it in place; otherwise wrap as `{"payload": original}`. Then
// inject id / tenant / type_id / version / embedding.
func augmentPayload(d searchabstraction.IndexDoc) ([]byte, error) {
	obj := map[string]any{}
	if raw := bytes.TrimSpace([]byte(d.Payload)); len(raw) > 0 {
		if raw[0] == '{' {
			if err := json.Unmarshal(raw, &obj); err != nil {
				return nil, err
			}
		} else {
			var inner any
			if err := json.Unmarshal(raw, &inner); err != nil {
				return nil, err
			}
			obj = map[string]any{"payload": inner}
		}
	}
	obj["id"] = string(d.ID)
	obj["tenant"] = string(d.Tenant)
	obj["type_id"] = string(d.TypeID)
	obj["version"] = d.Version
	if len(d.Embedding) > 0 {
		obj["embedding"] = d.Embedding
	}
	return json.Marshal(obj)
}

func (b *Backend) applyAuth(req *http.Request) {
	if b.authHeader != "" {
		req.Header.Set("Authorization", b.authHeader)
	}
}

func (b *Backend) sendJSON(ctx context.Context, method, u string, body any) (*http.Response, error) {
	var rdr io.Reader
	if body != nil {
		buf, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		rdr = bytes.NewReader(buf)
	}
	req, err := http.NewRequestWithContext(ctx, method, u, rdr)
	if err != nil {
		return nil, err
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	b.applyAuth(req)
	return b.http.Do(req)
}

// ─── SearchBackend impl ────────────────────────────────────────────────

// Search runs a tenant-scoped lexical query. Targets the concrete
// index when type_id is set, otherwise the tenant-wide alias.
func (b *Backend) Search(
	ctx context.Context,
	query searchabstraction.SearchQuery,
	_ repos.ReadConsistency,
) (repos.PagedResult[searchabstraction.SearchHit], error) {
	target := fmt.Sprintf("of-%s-*", sanitizeIndex(string(query.Tenant)))
	if query.TypeID != nil {
		target = b.indexName(query.Tenant, *query.TypeID)
	}
	u := fmt.Sprintf("%s/%s/_search", b.endpoint, target)

	filter := []any{map[string]any{"term": map[string]string{"tenant": string(query.Tenant)}}}
	if query.TypeID != nil {
		filter = append(filter, map[string]any{"term": map[string]string{"type_id": string(*query.TypeID)}})
	}
	for k, v := range query.Filters {
		filter = append(filter, map[string]any{"term": map[string]string{k: v}})
	}

	must := []any{}
	if query.Q != nil && *query.Q != "" {
		must = append(must, map[string]any{"query_string": map[string]string{"query": *query.Q}})
	}
	if len(must) == 0 {
		must = append(must, map[string]any{"match_all": map[string]any{}})
	}

	size := uint32(1)
	if query.Page.Size > 1 {
		size = query.Page.Size
	}
	body := map[string]any{
		"size":  size,
		"query": map[string]any{"bool": map[string]any{"must": must, "filter": filter}},
	}

	empty := repos.PagedResult[searchabstraction.SearchHit]{}
	resp, err := b.sendJSON(ctx, http.MethodPost, u, body)
	if err != nil {
		return empty, repos.Backend("opensearch search send: " + err.Error())
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNotFound {
		return repos.PagedResult[searchabstraction.SearchHit]{Items: []searchabstraction.SearchHit{}}, nil
	}
	if err := mapStatus(resp.StatusCode, "search"); err != nil {
		return empty, err
	}
	var v map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
		return empty, repos.Backend("opensearch search decode: " + err.Error())
	}
	return repos.PagedResult[searchabstraction.SearchHit]{Items: parseHits(v)}, nil
}

// Index PUTs a single document with `version_type=external`. HTTP
// 409 (stale external version) is silently swallowed.
func (b *Backend) Index(ctx context.Context, doc searchabstraction.IndexDoc) error {
	index := b.indexName(doc.Tenant, doc.TypeID)
	u := b.docURL(index, doc.ID)

	payload, err := augmentPayload(doc)
	if err != nil {
		return repos.Backend("opensearch index marshal: " + err.Error())
	}
	q := url.Values{}
	q.Set("version", fmt.Sprintf("%d", doc.Version))
	q.Set("version_type", "external")
	u = u + "?" + q.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodPut, u, bytes.NewReader(payload))
	if err != nil {
		return repos.Backend("opensearch index req: " + err.Error())
	}
	req.Header.Set("Content-Type", "application/json")
	b.applyAuth(req)
	resp, err := b.http.Do(req)
	if err != nil {
		return repos.Backend("opensearch index send: " + err.Error())
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusConflict {
		return nil
	}
	return mapStatus(resp.StatusCode, "index")
}

// Delete issues a tenant-wide _delete_by_query because the trait
// does not expose type_id at delete-time. Returns whether the
// document was present.
func (b *Backend) Delete(
	ctx context.Context,
	tenant repos.TenantId,
	id repos.ObjectId,
) (bool, error) {
	target := fmt.Sprintf("of-%s-*", sanitizeIndex(string(tenant)))
	u := fmt.Sprintf("%s/%s/_delete_by_query", b.endpoint, target)
	body := map[string]any{
		"query": map[string]any{
			"bool": map[string]any{
				"filter": []any{
					map[string]any{"term": map[string]string{"tenant": string(tenant)}},
					map[string]any{"term": map[string]string{"id": string(id)}},
				},
			},
		},
	}
	resp, err := b.sendJSON(ctx, http.MethodPost, u, body)
	if err != nil {
		return false, repos.Backend("opensearch delete send: " + err.Error())
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNotFound {
		return false, nil
	}
	if err := mapStatus(resp.StatusCode, "delete"); err != nil {
		return false, err
	}
	var v struct {
		Deleted uint64 `json:"deleted"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
		return false, repos.Backend("opensearch delete decode: " + err.Error())
	}
	return v.Deleted > 0, nil
}

// SearchVector runs an OpenSearch knn query. Filters are applied
// before the kNN search via the boolean `filter` clause.
func (b *Backend) SearchVector(
	ctx context.Context,
	query searchabstraction.VectorQuery,
	_ repos.ReadConsistency,
) ([]searchabstraction.SearchHit, error) {
	target := fmt.Sprintf("of-%s-*", sanitizeIndex(string(query.Tenant)))
	if query.TypeID != nil {
		target = b.indexName(query.Tenant, *query.TypeID)
	}
	u := fmt.Sprintf("%s/%s/_search", b.endpoint, target)

	filter := []any{map[string]any{"term": map[string]string{"tenant": string(query.Tenant)}}}
	if query.TypeID != nil {
		filter = append(filter, map[string]any{"term": map[string]string{"type_id": string(*query.TypeID)}})
	}
	for k, v := range query.Filters {
		filter = append(filter, map[string]any{"term": map[string]string{k: v}})
	}

	k := uint32(1)
	if query.K > 1 {
		k = query.K
	}
	body := map[string]any{
		"size": k,
		"query": map[string]any{
			"knn": map[string]any{
				"embedding": map[string]any{
					"vector": query.Embedding,
					"k":      k,
					"filter": map[string]any{"bool": map[string]any{"filter": filter}},
				},
			},
		},
	}
	resp, err := b.sendJSON(ctx, http.MethodPost, u, body)
	if err != nil {
		return nil, repos.Backend("opensearch vector send: " + err.Error())
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNotFound {
		return []searchabstraction.SearchHit{}, nil
	}
	if err := mapStatus(resp.StatusCode, "search_vector"); err != nil {
		return nil, err
	}
	var v map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
		return nil, repos.Backend("opensearch vector decode: " + err.Error())
	}
	return parseHits(v), nil
}

// BulkIndex writes a single _bulk request with one
// `{"index": …}` action per document plus the augmented payload. A
// per-document HTTP status of 200..=299 or 409 (stale write) counts
// as success — anything else is collected into BulkOutcome.Failed.
func (b *Backend) BulkIndex(
	ctx context.Context,
	docs []searchabstraction.IndexDoc,
) (searchabstraction.BulkOutcome, error) {
	out := searchabstraction.BulkOutcome{Failed: []searchabstraction.BulkFail{}}
	if len(docs) == 0 {
		return out, nil
	}
	var body bytes.Buffer
	for _, d := range docs {
		header := map[string]any{
			"index": map[string]any{
				"_index":       b.indexName(d.Tenant, d.TypeID),
				"_id":          string(d.ID),
				"version":      d.Version,
				"version_type": "external",
			},
		}
		hb, err := json.Marshal(header)
		if err != nil {
			return out, repos.Backend("opensearch bulk header marshal: " + err.Error())
		}
		body.Write(hb)
		body.WriteByte('\n')
		pb, err := augmentPayload(d)
		if err != nil {
			return out, repos.Backend("opensearch bulk payload marshal: " + err.Error())
		}
		body.Write(pb)
		body.WriteByte('\n')
	}

	u := fmt.Sprintf("%s/_bulk", b.endpoint)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u, &body)
	if err != nil {
		return out, repos.Backend("opensearch bulk req: " + err.Error())
	}
	req.Header.Set("Content-Type", "application/x-ndjson")
	b.applyAuth(req)
	resp, err := b.http.Do(req)
	if err != nil {
		return out, repos.Backend("opensearch bulk send: " + err.Error())
	}
	defer resp.Body.Close()
	if err := mapStatus(resp.StatusCode, "bulk_index"); err != nil {
		return out, err
	}
	var v struct {
		Items []map[string]json.RawMessage `json:"items"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
		return out, repos.Backend("opensearch bulk decode: " + err.Error())
	}
	for i, entry := range v.Items {
		raw, ok := entry["index"]
		if !ok {
			raw = entry["create"]
		}
		var id repos.ObjectId
		if i < len(docs) {
			id = docs[i].ID
		}
		if len(raw) == 0 {
			out.Failed = append(out.Failed, searchabstraction.BulkFail{ID: id, Reason: "missing status"})
			continue
		}
		var item struct {
			Status uint64 `json:"status"`
		}
		if err := json.Unmarshal(raw, &item); err != nil || item.Status == 0 {
			out.Failed = append(out.Failed, searchabstraction.BulkFail{ID: id, Reason: "missing status"})
			continue
		}
		switch {
		case item.Status >= 200 && item.Status < 300, item.Status == 409:
			out.Indexed++
		default:
			out.Failed = append(out.Failed, searchabstraction.BulkFail{
				ID: id, Reason: fmt.Sprintf("status %d", item.Status),
			})
		}
	}
	return out, nil
}

// parseHits decodes the OpenSearch `hits.hits[]` array. Mirrors
// fn parse_hits in the Rust impl.
func parseHits(v map[string]any) []searchabstraction.SearchHit {
	hitsRaw, ok := v["hits"].(map[string]any)
	if !ok {
		return []searchabstraction.SearchHit{}
	}
	arr, ok := hitsRaw["hits"].([]any)
	if !ok {
		return []searchabstraction.SearchHit{}
	}
	out := make([]searchabstraction.SearchHit, 0, len(arr))
	for _, raw := range arr {
		h, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		src, ok := h["_source"].(map[string]any)
		if !ok {
			continue
		}
		id, _ := src["id"].(string)
		if id == "" {
			id, _ = h["_id"].(string)
		}
		if id == "" {
			continue
		}
		typeID, _ := src["type_id"].(string)
		score := float32(0)
		if f, ok := h["_score"].(float64); ok {
			score = float32(f)
		}
		snippet, err := json.Marshal(src)
		if err != nil {
			snippet = nil
		}
		out = append(out, searchabstraction.SearchHit{
			ID:      repos.ObjectId(id),
			TypeID:  repos.TypeId(typeID),
			Score:   score,
			Snippet: snippet,
		})
	}
	return out
}
